### ISSUE TITLE
How to patch libraries that make copy of `requests.Session()`

### ISSUE THREAD

**Issue by glensc** [reclosedev/requests-cache#260]
Problem:

- https://github.com/pkkid/python-plexapi/blob/4.5.2/plexapi/server.py#L107

```py
    def __init__(self, baseurl=None, token=None, session=None, timeout=None):
...
        self._session = session or requests.Session()
```

It saves a copy of the requests object, so, therefore `requests_cache` can't do monkey-patch on/off.

What I want to do is make a decorator for methods in a class that does `with requests_cache.enabled(trakt_cache, expire_after=expire_after)`:
- https://github.com/Taxel/PlexTraktSync/pull/295/files

See my research:
- https://github.com/Taxel/PlexTraktSync/issues/307#issuecomment-830227300

**Comment by JWCook** [comment]
I would recommend [using a `CachedSession` object instead of `install_cache`](https://requests-cache.readthedocs.io/en/v0.6.3/user_guide.html#general-usage). The monkey-patching approach is convenient for the simplest use cases, but for anything beyond that, using `CachedSession` will usually result in cleaner code.

If you want to cache some requests but not others, there are a few different options besides `requests_cache.enabled`/`disabled`:
* Use a `CachedSession` for cached requests, and a regular `requests.Session` for non-cached requests
* Use a `CachedSession` for all requests, and wrap with [CachedSession.cache_disabled()](https://requests-cache.readthedocs.io/en/v0.6.3/api.html#requests_cache.session.CachedSession.cache_disabled) for non-cached requests
* Use [URL glob patterns](https://requests-cache.readthedocs.io/en/v0.6.3/user_guide.html#url-patterns) to define the requests you want to cache

You could still potentially make that work with `install_cache`, but it could get fairly messy, depending on how requests are being sent. If you can link me to an example method you would want to decorate with `@http_cache`, I might be able to come up with another alternative.

There are also some examples of using requests-cache with other libraries that modify/copy/subclass `requests.Session` here, which might give you some ideas:
https://requests-cache.readthedocs.io/en/v0.6.3/advanced_usage.html#library-compatibility

**Comment by glensc** [comment]
Currently, there's `@nocache` decorator, which will just wrap the `install_cache` context:
- https://github.com/Taxel/PlexTraktSync/blob/0.9.15/plex_trakt_sync/plex_api.py#L157-L159
- https://github.com/Taxel/PlexTraktSync/blob/0.9.15/plex_trakt_sync/decorators/nocache.py

**Comment by glensc** [comment]
I was thinking if I could tell Plex constructor `session=CachedSession`, and in decorator configure that session on/off, expiry time, that would satisfy the goal.

**Comment by JWCook** [comment]
Since your decorator has access to the `self` argument, will `self` be either a `PlexServer` or something that has access to an instance of `PlexServer`? If so, you could do something like:

```python
def http_cache(method, expire_after=None):
    @wraps(method)
    def inner(self, *args, **kwargs):
        # Replace self._session with however else you might access the session object from here
        with self._session.request_expire_after(expire_after):
            return method(self, *args, **kwargs)

    return inner
```

And your `@nocache` could look like:
```python
def nocache(method):
    @wraps(method)
    def inner(self, *args, **kwargs):
        with self._session.cache_disabled():
            return method(self, *args, **kwargs)

    return inner
```

**Comment by glensc** [comment]
@JWCook yes, it has access to the main plex server instance, but subclasses make more weird assignments, so might get too complicated if need to intercept those too:
- https://github.com/pkkid/python-plexapi/blob/4.5.2/plexapi/server.py#L503

but I think your suggestion will work, as it's the same object:

```py
with self.plex._session.cache_disabled():
	pass
```

altho that code is ugly, accessing private methods. so access the session passed to PlexServer construct instead:

```py
cache = CachedSession()
plex = PlexServer(session=cache)
...

def nocache(method):
    @wraps(method)
    def inner(self, *args, **kwargs):
        with cache.cache_disabled():
            return method(self, *args, **kwargs)

    return inner
```

**Comment by JWCook** [comment]
Yeah, that looks better. This should work as long as there's only one session object, and only one thread using it.

**Comment by JWCook** [comment]
I'm going to close this now since it looks like this is solved, but let me know if you have any more issues with this.
